### ISSUE TITLE
ARCH: Metanodes

### ISSUE THREAD

**Issue by CJ-Wright** [xpdAcq/SHED#112]
I've been thinking about metanodes (as discussed with @tacaswell @danielballan @stuartcampbell and @jrmlhermitte @sbillinge ).

My understanding (which may be flawed) is that:
1. At the boundary of the metanode going in we will extract data from whatever input streams we need. This extraction will go from the event model documents to python base types + numpy + whatever other types we need to traffic in (but not event model). 
1. We will do work on the data (potentially via streamz or something else, this is a bit un-clear to me). 
1. Finally we will re-package the data on the way out turning it back into event model.

I am a little uncertain about the re-packaging. How does this work?
1. We can't use data inside the metanode, since it is devoid of any info about the start and stop of an experiment.
1. It may be difficult to use the input data (before the translation) to determine when to issue a stop/start on the way out as we may ingest data much faster than we process it.

Does anyone have ideas on this?

**Comment by jrmlhermitte** [comment]
We also discussed that we need more use cases. I can try to provide one. It may not be the best implementation but it is allowing us to have a better idea on how we want to process our data, and start analyzing pain points. (I'll explicitly write out where `SHED` could be useful)

There are three steps we use:
1. Use a callback to convert from event model to data done in [this line](https://github.com/CFN-softbio/SciStreams/blob/master/SciStreams/startup/run_stream_test.py#L87), the callback object is [this object](https://github.com/CFN-softbio/SciStreams/blob/master/SciStreams/callbacks/core.py#L26).
2. Run analysis (the rest of the script). Assumes the data element is a dictionary, which contains various stuff inside (metadata or numpy arrays etc). This I guess would be the metanode stuff that `EventStream` would not care about.
3. Create various sinks that sink the data, whatever it is back to the event model [here](https://github.com/CFN-softbio/SciStreams/blob/master/SciStreams/startup/run_stream_test.py#L285). This event stream then goes to various callbacks like one custom written callback to store a matplotlib plot as lossy jpeg [here](https://github.com/CFN-softbio/SciStreams/blob/master/SciStreams/startup/run_stream_test.py#L321).

The stream itself does a few things which are outputted as event streams again from [this line](https://github.com/CFN-softbio/SciStreams/blob/master/SciStreams/startup/run_stream_test.py#L285) to [this line](https://github.com/CFN-softbio/SciStreams/blob/master/SciStreams/startup/run_stream_test.py#L296):
1. Grab an image (which can get plotted etc)
2. Get the polar projected q-phi map of image
3. Get the 1D S(q) curve
4. Get the peak indices of the 1D S(q) curve through peak finding
5. Get the line cuts of the polar projected map at each of these found peaks (which right now are just eventually plotted as a few graphs in matplotlib)
6. Get the angular correlation at these found peaks plotted as 1D curves
7. Get the 2 dimensional angular correlation map
8. Get the tag of an image (which right now gets saved to an hdf5 file and xml file) which was obtained by running inference on a pre-trained convolutional neural networks. Unfortunately, we keep the code for the CNN out of this code base to respect our colleagues at CSI.

We will have some use cases for GISAXS soon and some better applications using dask. Currently We only use dask when computing the qphimaps [here](https://github.com/CFN-softbio/SciStreams/blob/master/SciStreams/streams/XS_Streams.py#L356). It's messy but it works, and the data is properly cached. (The caching is done by defining how to hash the object that is passed to the functions. We're pickling arbitrary objects, but it's a compromise. We lose efficiency as opposed to just sending numpy arrays, but we gain by allowing scientists to define their own objects)


Where I see `EventStream` can play a role is at the beginning of 1 or end of 3. We definitely find that at step 2, making the assumption that things are homogeneous are extremely useful and more practical.

Some examples for possible uses of `shed` (and please add to this list if you find more :-) ):
- One could imagine the data coming in could arrive from multiple sources. An object that was able to correctly multiplex incoming data to correct analysis functions (or streams or anything) would be very useful.
- Another potential problem one could imagine is one has two streams coming in: temperature, and images, acquired at different times. It would be *extremely* useful if we could develop some method to take these two streams together and provide synchronized data. For example, if the temperature is taken at higher frequency than the images, then interpolate the temperature to the images. I believe that this is also something @danielballan has been working on with datamuxer (Dan please correct me)
- One may again want to do the same thing with the data outputted from the analysis functions

All in all, I think what our use cases are showing us (and it may not be the best use case though...) is that shed may be too complex for running analysis on well defined data (like the different analysis we're running I mentioned above) but could be **very** useful as a data normalizer and data muxxer. That might be a good place to focus on.


Also, it might be good to summarize use cases somewhere I apologize for the long response here.
Perhaps in some form of documentation or just some wiki we can keep track of for now?

**Comment by tacaswell** [comment]
> We can't use data inside the metanode, since it is devoid of any info about the start and stop of an experiment.

Why not?  It came across the edge in so you have it.   You can keep what ever notes / persistent state internally that you need.

> It may be difficult to use the input data (before the translation) to determine when to issue a stop/start on the way out as we may ingest data much faster than we process it.

At some point these _have_ to balance or your backlog will grow forever.   There will have to be some form of back-pressure someplace which can presumably be used to help manage this.  

Both of these questions are, I think, implementation specific but require some sort of communication between the input and output edge.  This could be passing a sentinel through to separate runs (maybe containing metadata for the start document),  some sort of side-band between the source and the sink, or slightly structured data going through the pipeline, etc .

Put another way, the primary job of the metanode is to do this book keeping!

> Another potential problem one could imagine is one has two streams coming in: temperature, and images, acquired at different times.

datamuxer is more aimed at re-sampling data post-fact.  Down sampling is pretty straight forward (assuming you know the time box of the slow data) and up-sampling post-fact is pretty straight forward.  Much of this (for scalars) is already implemented in pandas.

**Comment by danielballan** [comment]
> For example, if the temperature is taken at higher frequency than the images, then interpolate the temperature to the images. I believe that this is also something @danielballan has been working on with datamuxer (Dan please correct me)

Right. Our datamuxer library implements this by using pandas because pandas has great resampling functionality. It could now be improved using streaming DataFrames.

**Comment by danielballan** [comment]
...and I'll add that my _new_ work on this hasn't gone beyond the experimental phase. I'd like to get a couple solid days to focus on it.

**Comment by CJ-Wright** [comment]
@tacaswell admittedly structured data is what is going through the pipeline right now, but we could try to reduce the structure some (just include info about the start uids going in).

One possibility would to have some sort of dedicated metadata object (which is really just a dict or something) that Streamz can understand not to process but just to pass down with some info. We'd have to carve out some space inside all the nodes to make certain that they handled the data properly. The translator nodes could add the uids to the metadata and when they hit boundary they build a new document stream based off of this data. They could check if certain pieces of the metadata have changed (eg the raw data stream uid) and if so then they break off a new stream of documents.

**Comment by CJ-Wright** [comment]
See https://github.com/mrocklin/streamz/issues/110

**Comment by CJ-Wright** [comment]
I think this issue has been fixed. The translation nodes properly sideband start and stop document information from `FromEventStream` to `ToEventStream`. This allows the implementation of metanodes however needed. Eg one could create some mega data processing class (not advised, but ...) which accepts a whole slew of raw data. This class is instantiated and put inside a map node. Data is passed in, and data comes out. `ToEventStream` re-packages this in the event model. On the flip side one could have a very intricate `rapdiz` pipeline which describes the analysis and that works too.

`rapidz` now has nice tools for taking chunks, functions which describe a part of the pipeline, and stringing them together to make a full pipeline (essentially solving the graph). These chunks might map nicely to metanodes.
